src/frontend/src/expr/mod.rs

```diff
@@ -750,9 +750,31 @@ impl ExprImpl {
             true
         } else if let ExprImpl::FunctionCall(f) = self {
             match f.get_expr_type() {
-                ExprType::Add | ExprType::Subtract => {
+                ty @ (ExprType::Add
+                | ExprType::Subtract
+                | ExprType::AtTimeZone
+                | ExprType::CastWithTimeZone) => {
                     let (_, lhs, rhs) = f.clone().decompose_as_binary();
-                    lhs.is_now_offset() && rhs.is_const()
+                    if matches!(ty, ExprType::AtTimeZone | ExprType::CastWithTimeZone)
+                        && rhs
+                            .as_literal()
+                            .and_then(|literal| literal.get_data().as_ref())
+                            .map_or(true, |time_zone| *time_zone != String::from("UTC").into())
+                    {
+                        false
+                    } else {
+                        lhs.is_now_offset() && rhs.is_const()
```

Comment:
FYI not an issue of this PR, but `is_const` is not enough. The interval const cannot be variable length (eg. `'1' month`):

```
test=# select v, v + interval '1' month from generate_series('2023-01-27'::timestamp, '2023-02-01'::timestamp, interval '12' hour) as t(v);
          v          |      ?column?       
---------------------+---------------------
 2023-01-27 00:00:00 | 2023-02-27 00:00:00
 2023-01-27 12:00:00 | 2023-02-27 12:00:00
 2023-01-28 00:00:00 | 2023-02-28 00:00:00
 2023-01-28 12:00:00 | 2023-02-28 12:00:00
 2023-01-29 00:00:00 | 2023-02-28 00:00:00
 2023-01-29 12:00:00 | 2023-02-28 12:00:00
 2023-01-30 00:00:00 | 2023-02-28 00:00:00
 2023-01-30 12:00:00 | 2023-02-28 12:00:00
 2023-01-31 00:00:00 | 2023-02-28 00:00:00
 2023-01-31 12:00:00 | 2023-02-28 12:00:00
 2023-02-01 00:00:00 | 2023-03-01 00:00:00
(11 rows)
```

For `timestamptz`, the length of `'1' day` is also variable. To be fully correct, we would need to evaluate the const and check its `months` and `days` are both zero.